LoggingPlugin: Support to customize log_file from hook

mitzkia opened the pull request:
First of all thanks to @thisch who allowed to make this PR possible.

Without this patch there was only one static way to customize log_file path: --log-file starter option.
This patch allows to customize log_file value also from hooks, which helps to create log-files dynamically even for each testcases.

blueyed commented:
> blueyed added some commits 5 hours ago

Looks like this is not really based on features, but on master (or similar).

Please rebase it on features, and squash your commits into a single one.

blueyed commented on doc/en/logging.rst:
```diff
@@ -198,6 +198,16 @@ option names are:
 * ``log_file_format``
 * ``log_file_date_format``
 
+You can call ``set_log_filename()`` to customize log_file path dinamically::
```
```suggestion
You can call ``set_log_filename()`` to customize the log_file path dynamically::
```

mitzkia replied:
Thank you for the review. The typo is fixed, and also commits are squashed and rebased.

mitzkia commented:
Question: 
For now there is a logic in `set_log_filename()` where we creates dir (if it is not exist) for our logfile. But we uses `os.path..` form which is not compatible with Path() objects (I prefer this type).
I can fix this with checking type of fname in set_log_filename() and creating dir in proper way.
Would it be fine?

RonnyPfannschmidt commented:
i believe for consistency with old python we need to the  function to be called `set_log_path`, and cast it to `Path objects

in case we get a relative path, we need to choose what we use as base as well,

nicoddemus commented:
> in case we get a relative path, we need to choose what we use as base as well,

`config.rootdir` seems like a good candidate for that.

About the parameter, it should just be `str`, no need to require it to be `Path` objects.

RonnyPfannschmidt commented:
@nicoddemus the parameter must support being cast to a Path, it must not be a Path

mitzkia commented:
I have added some fixes:
- rename from set_log_filename() -> set_log_path() (also in docs)
I forgot that logging.FileHandler() can deal with only `str` instead of Path(), so I have added a check for this, and cast Path() objects to str. 

Next I will add checking for abs/rel path and fix the base path if needed.

nicoddemus commented on src/_pytest/logging.py:
```diff
@@ -468,6 +469,16 @@ def _setup_cli_logging(self):
         )
 
+    def set_log_path(self, fname):
+        if isinstance(fname, pathlib.PurePath):
```
I think just `fname = Path(fname)` is enough instead of this check. 👍

mitzkia replied:
Sorry, did not understand why to cast vica versa? 
At the end we will need `str` type for `fname` to work correctly within `logging.FileHandler()`.
`fname` parameter can be `str`, or `pathlib` type. If it is `str` type everything is ok, `logging.FileHandler()` will work. But if it is `pathlib` type I would cast it to `str` at beginning of this function.
Please correct me if I am wrong. Thanks

mitzkia replied:
Sorry, Just read: https://docs.python.org/3/library/logging.handlers.html#logging.FileHandler, 
From 3.6 it supports Path() also. But I have an exception if filename in `pathlib` type

mitzkia replied:
Maybe for compatibility reasons it is ok if the type is str

RonnyPfannschmidt replied:
we should use pathlib and cast to str until we drop suport for python < 3.6

mitzkia replied:
Thanks, fixed

nicoddemus commented on src/_pytest/logging.py:
```diff
@@ -468,6 +469,16 @@ def _setup_cli_logging(self):
+        if isinstance(fname, pathlib.PurePath):
+            fname = str(fname)
+
+        if not os.path.exists(os.path.dirname(fname)):
```
Following my advice above, it is just a matter of changing this to `fname.parent.mkdir(exist_ok=True, parents=True)`

mitzkia replied:
Thanks, fixed

nicoddemus commented:
Also thanks a ton to @mitzkia for the contribution! 🤗

thisch commented:
Thank you very much @mitzkia for working on this!

@nicoddemus mentioned in https://github.com/pytest-dev/pytest/issues/4707#issuecomment-460810153 that this feature should be marked as experimental. Not sure how this can be done. Maybe @nicoddemus  can comment on what needs to be done to mark this feature as experimental.

Please do not forget to add an unittest to this PR.

nicoddemus commented on doc/en/logging.rst:
```diff
@@ -198,6 +198,16 @@ option names are:
 * ``log_file_format``
 * ``log_file_date_format``
 
+You can call ``set_log_path()`` to customize the log_file path dynamically::
```
I think it is enough here to mention that this feature is experimental.

mitzkia replied:
Thank you, fixed

thisch replied:
I think we should add the example hook (that was part of  https://github.com/pytest-dev/pytest/commit/77b0dd3f95693be67528f7afb449eca9a79d026a once this feature is no longer experimental.

mitzkia replied:
Thanks @thisch. I am agreed with you. I will take a note for myself about this. Is there a best practice to not forgot this kind of future changes?

thisch replied:
I would simply create a ticket for this.

nicoddemus commented on src/_pytest/logging.py:
```diff
@@ -468,6 +468,20 @@ def _setup_cli_logging(self):
             log_cli_handler, formatter=log_cli_formatter, level=log_cli_level
         )
 
+    def set_log_path(self, fname):
```
Please write a docstring here given that this is now public. Also mention that this method is still experimental. 👍

mitzkia replied:
Thank you, fixed.

mitzkia commented:
I have fixed latest review notes and also added working unit test.

nicoddemus commented on testing/logging/test_reporting.py:
```diff
@@ -1002,3 +1002,49 @@ def pytest_sessionfinish(session, exitstatus):
         assert "sessionfinish" in contents
+
+def test_log_set_path(testdir):
+    from pathlib2 import Path
```
This needs to be `from _pytest.pathlib import Path` so it works for all Python versions we support.

mitzkia replied:
Sorry, about that. Previously I already used that way (as you mentioned), just forgot. Soon I will fix.

mitzkia replied:
Updated

nicoddemus commented on testing/logging/test_reporting.py:
```diff
@@ -1002,3 +1002,49 @@ def pytest_sessionfinish(session, exitstatus):
+    testdir.makeconftest(
+        """
+            import pytest
+            from pathlib2 import Path
```
This needs to be `from _pytest.pathlib import Path` so it works for all Python versions we support.

mitzkia replied:
Thanks the review notes, I will fix this also

mitzkia replied:
Updated

mitzkia commented:
I know about TravisCI failing, soon I will fix it.

RonnyPfannschmidt commented:
@mitzkia thanks for the followup, you are doing great, please sort things out at a pace you are most comfortable with

mitzkia commented:
Sorry, but I can not figure out what could be the last CI error on Windows, I have already tried many possible fixes, but none of them worked. Unfortunately I have not got such Windows environment to try it.

Thanks a lot for the help.

nicoddemus commented:
@mitzkia sure thing, the problem is that you are inlining the full path name into the source code:

```
  File "C:\Users\appveyor\AppData\Local\Temp\1\pytest-of-appveyor\pytest-0\popen-gw0\test_log_set_path0\conftest.py", line 7
    report_file = os.path.join("C:\Users\appveyor\AppData\Local\Temp\1\pytest-of-appveyor\pytest-0\popen-gw0\test_log_set_path0", item._request.node.name)
```

And writing `report_file = os.path.join("C:\User...` (without escaping) produces that error. I've changed to use `repr()` while writing the source (see my previous commit).

mitzkia commented:
@nicoddemus Thank you very much for the help and explanation.

mitzkia commented:
I have updated the branch.
